File: pipeline/style_parity.go
package main

// style-parity, ported from tools/style-parity.mjs — COMPUTED STYLE parity
// between the React oracle and the shadless fixture, per contract component.
// Pair by data-slot sequence, compare getComputedStyle over a property list;
// deterministic, computed values rounded to 2dp.

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

const styleParityDir = "tools/contracts/out"

var styleParityProps = []string{
	"color", "background-color", "border-color", "border-top-width", "border-radius",
	"display", "position", "flex-direction", "align-items", "justify-content",
	"gap", "padding-top", "padding-right", "padding-bottom", "padding-left",
	"margin-top", "margin-right", "margin-bottom", "margin-left",
	"width", "height", "min-width", "max-width", "font-size", "font-weight",
	"line-height", "text-align", "opacity", "z-index", "overflow", "visibility",
	"box-shadow", "transform", "inset", "flex-wrap", "grid-template-columns",
}

// runtime-measured / animation noise both sides carry differently
var styleParitySkip = map[string]bool{
	"transform": true, "transition": true, "transition-duration": true, "animation": true,
}

// Measurements were being taken MID-ANIMATION: freeze both sides first.
const styleParityFreeze = `*, *::before, *::after {
  transition: none !important; animation: none !important;
  animation-duration: 0s !important; transition-duration: 0s !important;
}`

// Harness shell, identical on both sides (neither convention is the component).
const styleParityShell = "padding:0;margin:0;color:var(--foreground);background:var(--background)"

const styleParityCollect = `(props) => {
  const out = []
  const walk = (el) => {
    if (el.tagName === "SCRIPT" || el.tagName === "TEMPLATE") return
    if (el.hasAttribute("data-slot")) {
      const cs = getComputedStyle(el)
      const style = {}
      for (const p of props) style[p] = cs.getPropertyValue(p)
      out.push({ slot: el.getAttribute("data-slot"), tag: el.tagName, style })
    }
    for (const c of el.children) walk(c)
  }
  for (const root of document.body.children) walk(root)
  return out
}`

var reStyleCalc = regexp.MustCompile(`calc\([^)]*\)`)

func styleParityNorm(v string) string {
	if v == "" {
		return v
	}
	v = reParityNum.ReplaceAllStringFunc(v, func(n string) string {
		r := parseFloat2dp(n)
		if r == 0 {
			return "0" // Object.is(-0) guard
		}
		return jsNumberString(r)
	})
	v = reParityOklab.ReplaceAllString(v, "oklch($1 0 0)")
	return reStyleCalc.ReplaceAllString(v, "calc(…)")
}

type spCellEnt struct {
	component string
	key       string
	prop      string
	oracle    string
	shadless  string
}

func runStyleParity(strict, record bool) int {
	const baselinePath = "gates/style-parity-baseline.json"

	shell, err := startBrowserShell()
	if err != nil {
		fmt.Fprintln(os.Stderr, "style-parity:", err)
		return 1
	}
	defer shell.close()
	if err := shell.launch(); err != nil {
		fmt.Fprintln(os.Stderr, "style-parity:", err)
		return 1
	}
	page, _ := shell.newPage(false)
	page.routeAbortExternal()

	var cells []spCellEnt
	var harnessErrors []string

	ents, _ := os.ReadDir("tools/contracts/components")
	var names []string
	for _, e := range ents {
		if strings.HasSuffix(e.Name(), ".mjs") {
			names = append(names, strings.TrimSuffix(e.Name(), ".mjs"))
		}
	}
	sort.Strings(names)

	compared, components := 0, 0
	variants := []struct {
		id   string
		dark bool
		dir  string
	}{
		{"light@ltr", false, "ltr"}, {"dark@ltr", true, "ltr"},
		{"light@rtl", false, "rtl"}, {"dark@rtl", true, "rtl"},
	}

	collectMatrix := func() map[string][]map[string]any {
		out := map[string][]map[string]any{}
		for _, v := range variants {
			page.evaluateFnArg(`(v) => {
        document.documentElement.classList.toggle("dark", v.dark)
        document.documentElement.setAttribute("dir", v.dir)
      }`, map[string]any{"dark": v.dark, "dir": v.dir})
			page.waitForTimeout(60)
			res, err := page.evaluateFnArg(styleParityCollect, styleParityProps)
			if err != nil {
				return out
			}
			if arr, ok := res.([]any); ok {
				for _, e := range arr {
					if m, ok := e.(map[string]any); ok {
						out[v.id] = append(out[v.id], m)
					}
				}
			}
		}
		page.evaluate(`(function(){ document.documentElement.classList.remove("dark"); document.documentElement.setAttribute("dir","ltr") })()`)
		return out
	}
	freeze := func() {
		page.evaluateFnArg(`(shell) => { document.body.style.cssText += ";" + shell }`, styleParityShell)
		page.addStyleTag(styleParityFreeze)
		page.evaluate(`(function(){ document.getAnimations?.().forEach((a) => a.finish()) })()`)
		page.waitForTimeout(120)
	}

	for _, name := range names {
		dir := filepath.Join(styleParityDir, name)
		if !fileExists(filepath.Join(dir, "oracle.html")) || !fileExists(filepath.Join(dir, "shadless.html")) {
			continue
		}
		defRes, err := shell.call(map[string]any{
			"op":         "loadContractDef",
			"file":       "file://" + absOrDie(filepath.Join(cwd0(), "tools/contracts/components", name+".mjs")),
		})
		if err != nil {
			harnessErrors = append(harnessErrors, name+": harness error — "+firstLine(err.Error()))
			continue
		}
		var def struct {
			Open         string   `json:"open"`
			OpenShadless string   `json:"openShadless"`
			StyleIgnore  []string `json:"styleIgnore"`
		}
		b, _ := json.Marshal(defRes["def"])
		json.Unmarshal(b, &def)

		styleIgnore := map[string]bool{}
		for _, p := range def.StyleIgnore {
			styleIgnore[p] = true
		}
		for p := range styleParitySkip {
			styleIgnore[p] = true
		}
		var props []string
		for _, p := range styleParityProps {
			if !styleIgnore[p] {
				props = append(props, p)
			}
		}

		func() {
			defer func() {
				if r := recover(); r != nil {
					harnessErrors = append(harnessErrors, fmt.Sprintf("%s: harness error — %v", name, r))
				}
			}()
			// oracle side (styled by upstream's own oracle.css)
			absO, _ := filepath.Abs(filepath.Join(dir, "oracle.html"))
			if err := page.gotoURL("file://" + absO); err != nil {
				panic(err)
			}
			page.waitForTimeout(400)
			if def.Open != "" {
				if err := shell.callErr(map[string]any{"op": "driver", "pageId": page.id, "code": def.Open}); err != nil {
					panic(err)
				}
				page.waitForTimeout(400)
			}
			page.addStyleTagPath("build/gates/oracle.css")
			page.evaluate(`document.documentElement.classList.add("style-nova")`)
			freeze()
			oracleSides := collectMatrix()

			// shadless side (loads its own out.css via relative link)
			absS, _ := filepath.Abs(filepath.Join(dir, "shadless.html"))
			if err := page.gotoURL("file://" + absS); err != nil {
				panic(err)
			}
			page.addStyleTagPath("dist/out.css")
			page.waitForTimeout(400)
			open := def.OpenShadless
			if open == "" {
				open = def.Open
			}
			if open != "" {
				if err := shell.callErr(map[string]any{"op": "driver", "pageId": page.id, "code": open}); err != nil {
					panic(err)
				}
				page.waitForTimeout(400)
			}
			freeze()
			shadlessSides := collectMatrix()

			slotOf := func(e map[string]any) string { s, _ := e["slot"].(string); return s }
			for _, variant := range variants {
				seenA := map[string]int{}
				seenB := map[string]int{}
				keyOf := func(slot string, seen map[string]int) string {
					seen[slot]++
					return fmt.Sprintf("%s#%d", slot, seen[slot])
				}
				mapA := map[string]map[string]string{}
				var orderA []string
				for _, e := range oracleSides[variant.id] {
					k := keyOf(slotOf(e), seenA)
					style, _ := e["style"].(map[string]any)
					sm := map[string]string{}
					for p, v := range style {
						sm[p], _ = v.(string)
					}
					mapA[k] = sm
					orderA = append(orderA, k)
				}
				mapB := map[string]map[string]string{}
				var orderB []string
				for _, e := range shadlessSides[variant.id] {
					k := keyOf(slotOf(e), seenB)
					style, _ := e["style"].(map[string]any)
					sm := map[string]string{}
					for p, v := range style {
						sm[p], _ = v.(string)
					}
					mapB[k] = sm
					orderB = append(orderB, k)
				}
				push := func(k, prop, a, b string) {
					key := k
					if variant.id != "light@ltr" {
						key = k + "@" + variant.id
					}
					cells = append(cells, spCellEnt{name, key, prop, a, b})
				}
				for _, k := range orderA {
					a := mapA[k]
					b, ok := mapB[k]
					if !ok {
						push(k, "<presence>", "present", "missing")
						continue
					}
					for _, p := range props {
						va := styleParityNorm(a[p])
						vb := styleParityNorm(b[p])
						if va != vb {
							push(k, p, va, vb)
						}
					}
				}
				for _, k := range orderB {
					if _, ok := mapA[k]; !ok {
						push(k, "<presence>", "missing", "present")
					}
				}
				compared += len(mapA)
			}
			components++
		}()
	}

	if len(harnessErrors) > 0 {
		fmt.Fprintf(os.Stderr, "FAIL  style-parity (harness)\n  %s\n", strings.Join(harnessErrors, "\n  "))
		return 1
	}

	// `flaky` from the raw file (record must work pre-value format too)
	flaky := map[string]bool{}
	if fb, err := os.ReadFile(baselinePath); err == nil {
		var raw struct {
			Flaky []string `json:"flaky"`
		}
		json.Unmarshal(fb, &raw)
		for _, f := range raw.Flaky {
			flaky[f] = true
		}
	}

	var ratcheted []spCellEnt
	for _, c := range cells {
		if !flaky[c.component] {
			ratcheted = append(ratcheted, c)
		}
	}
	flakyCells := len(cells) - len(ratcheted)
	var actualCells []parityCell
	for _, c := range ratcheted {
		actualCells = append(actualCells, parityCell{c.component + "/" + c.key + "/" + c.prop, c.oracle, c.shadless})
	}
	actual, order := cellMap(actualCells)

	if record || !fileExists(baselinePath) {
		var fl []string
		for f := range flaky {
			fl = append(fl, f)
		}
		if err := writeParityBaseline(baselinePath,
			"Cells where the shadless fixture's computed style differs from the React oracle, "+
				"with the two values as recorded. This list may only shrink and the values are pinned; "+
				"see the ledger budget style-parity.dirty-cells.",
			fl, actual); err != nil {
			fmt.Fprintln(os.Stderr, "style-parity:", err)
			return 1
		}
		compSet := map[string]bool{}
		for _, c := range ratcheted {
			compSet[c.component] = true
		}
		fmt.Printf("style-parity: baseline recorded (%d cells across %d components, %d flaky components excluded)\n",
			len(actual), len(compSet), len(flaky))
		return 0
	}

	if strict && len(cells) > 0 {
		n := 10
		if len(cells) < n {
			n = len(cells)
		}
		var parts []string
		for _, c := range cells[:n] {
			parts = append(parts, c.component+"/"+c.key+"/"+c.prop+": oracle="+trunc60(c.oracle)+" shadless="+trunc60(c.shadless))
		}
		fmt.Fprintf(os.Stderr, "FAIL  style-parity --strict (%d differing cells)\n  %s\n", len(cells), strings.Join(parts, "\n  "))
		return 1
	}

	_, recorded, err := loadParityBaseline(baselinePath)
	if err != nil {
		fmt.Fprintln(os.Stderr, "style-parity:", err)
		return 1
	}
	d := diffParityBaseline(recorded, actual, order)
	if len(d.appeared) > 0 {
		n := 40
		if len(d.appeared) < n {
			n = len(d.appeared)
		}
		var parts []string
		for _, id := range d.appeared[:n] {
			parts = append(parts, id+": "+showCell(actual[id]))
		}
		tail := ""
		if len(d.appeared) > 40 {
			tail = fmt.Sprintf("\n  … +%d more", len(d.appeared)-40)
		}
		fmt.Fprintf(os.Stderr, "FAIL  style-parity (%d NEW differing cells vs the React oracle)\n  %s%s\n",
			len(d.appeared), strings.Join(parts, "\n  "), tail)
		return 1
	}
	if len(d.changed) > 0 {
		n := 20
		if len(d.changed) < n {
			n = len(d.changed)
		}
		var parts []string
		for _, c := range d.changed[:n] {
			parts = append(parts, showChange(c))
		}
		tail := ""
		if len(d.changed) > 20 {
			tail = fmt.Sprintf("\n  … +%d more", len(d.changed)-20)
		}
		fmt.Fprintf(os.Stderr, "FAIL  style-parity (%d recorded cells still differ from the oracle, but by a DIFFERENT amount than what was recorded — look at them again, then re-record)\n  %s%s\n\n  ./build/pipeline style-parity --record\n",
			len(d.changed), strings.Join(parts, "\n  "), tail)
		return 1
	}
	if len(d.fixed) > 0 {
		n := 12
		if len(d.fixed) < n {
			n = len(d.fixed)
		}
		tail := ""
		if len(d.fixed) > 12 {
			tail = fmt.Sprintf("\n  … +%d more", len(d.fixed)-12)
		}
		fmt.Fprintf(os.Stderr, "FAIL  style-parity (%d recorded cells no longer differ — record the win so the slack cannot be re-spent)\n  %s%s\n\n  ./build/pipeline style-parity --record && ./build/pipeline ledger --record\n",
			len(d.fixed), strings.Join(d.fixed[:n], "\n  "), tail)
		return 1
	}
	fmt.Printf("PASS  style-parity (%d components, %d elements compared, %d cells at the recorded baseline incl. their values, %d cells in %d flaky components excluded; --strict is the end state)\n",
		components, compared, len(actual), flakyCells, len(flaky))
	return 0
}

func cwd0() string {
	wd, _ := os.Getwd()
	return wd
}

func absOrDie(p string) string {
	abs, err := filepath.Abs(p)
	if err != nil {
		return p
	}
	return abs
}

File: pipeline/src/tools/style_parity.rs
//! Port of pipeline/style_parity.go — COMPUTED STYLE parity between the React
//! oracle and the shadless fixture, per contract component. Pair by data-slot
//! sequence, compare getComputedStyle over a property list; deterministic,
//! computed values rounded to 2dp.

use serde::Deserialize;
use std::path::Path;

use super::parity_baseline::{
    cell_map, diff_parity_baseline, load_parity_baseline, parity_cell, parity_norm_value,
    show_cell, show_change, write_parity_baseline,
};

const STYLE_PARITY_DIR: &str = "tools/contracts/out";
const STYLE_PARITY_BASELINE: &str = "gates/style-parity-baseline.json";

const STYLE_PARITY_PROPS: [&str; 36] = [
    "color", "background-color", "border-color", "border-top-width", "border-radius",
    "display", "position", "flex-direction", "align-items", "justify-content",
    "gap", "padding-top", "padding-right", "padding-bottom", "padding-left",
    "margin-top", "margin-right", "margin-bottom", "margin-left",
    "width", "height", "min-width", "max-width", "font-size", "font-weight",
    "line-height", "text-align", "opacity", "z-index", "overflow", "visibility",
    "box-shadow", "transform", "inset", "flex-wrap", "grid-template-columns",
];

// runtime-measured / animation noise both sides carry differently
const STYLE_PARITY_SKIP: [&str; 4] = ["transform", "transition", "transition-duration", "animation"];

// Measurements were being taken MID-ANIMATION: freeze both sides first.
const STYLE_PARITY_FREEZE: &str = r#"*, *::before, *::after {
  transition: none !important; animation: none !important;
  animation-duration: 0s !important; transition-duration: 0s !important;
}"#;

// Harness shell, identical on both sides (neither convention is the component).
const STYLE_PARITY_SHELL: &str = "padding:0;margin:0;color:var(--foreground);background:var(--background)";

const STYLE_PARITY_COLLECT: &str = r#"(props) => {
  const out = []
  const walk = (el) => {
    if (el.tagName === "SCRIPT" || el.tagName === "TEMPLATE") return
    if (el.hasAttribute("data-slot")) {
      const cs = getComputedStyle(el)
      const style = {}
      for (const p of props) style[p] = cs.getPropertyValue(p)
      out.push({ slot: el.getAttribute("data-slot"), tag: el.tagName, style })
    }
    for (const c of el.children) walk(c)
  }
  for (const root of document.body.children) walk(root)
  return out
}"#;

#[derive(Deserialize, Default)]
struct SpDef {
    #[serde(default)]
    open: String,
    #[serde(rename = "openShadless", default)]
    open_shadless: String,
    #[serde(rename = "styleIgnore", default)]
    style_ignore: Vec<String>,
}

struct SpCellEnt {
    component: String,
    key: String,
    prop: String,
    oracle: String,
    shadless: String,
}

pub fn run_style_parity(root: &Path, strict: bool, record: bool) -> i32 {
    let shell = match crate::oracle::browser_shell::BrowserShell::start() {
        Ok(s) => s,
        Err(e) => {
            eprintln!("style-parity: {}", e);
            return 1;
        }
    };
    if let Err(e) = shell.launch() {
        eprintln!("style-parity: {}", e);
        return 1;
    }
    let page = match shell.new_page(false) {
        Ok(p) => p,
        Err(e) => {
            eprintln!("style-parity: {}", e);
            return 1;
        }
    };
    let _ = page.route_abort_external();

    let mut cells: Vec<SpCellEnt> = Vec::new();
    let mut harness_errors: Vec<String> = Vec::new();

    let mut names: Vec<String> = Vec::new();
    if let Ok(ents) = std::fs::read_dir(root.join("tools/contracts/components")) {
        for e in ents.flatten() {
            let n = e.file_name().to_string_lossy().into_owned();
            if let Some(stem) = n.strip_suffix(".mjs") {
                names.push(stem.to_string());
            }
        }
    }
    names.sort();

    let (mut compared, mut components) = (0usize, 0usize);
    let variants: [(&str, bool, &str); 4] = [
        ("light@ltr", false, "ltr"),
        ("dark@ltr", true, "ltr"),
        ("light@rtl", false, "rtl"),
        ("dark@rtl", true, "rtl"),
    ];

    let collect_matrix = || -> std::collections::HashMap<String, Vec<serde_json::Value>> {
        let mut out: std::collections::HashMap<String, Vec<serde_json::Value>> =
            std::collections::HashMap::new();
        for v in variants {
            let _ = page.evaluate_fn_arg(
                r#"(v) => {
        document.documentElement.classList.toggle("dark", v.dark)
        document.documentElement.setAttribute("dir", v.dir)
      }"#,
                serde_json::json!({"dark": v.1, "dir": v.2}),
            );
            page.wait_for_timeout(60);
            let Ok(res) = page.evaluate_fn_arg(STYLE_PARITY_COLLECT, serde_json::json!(STYLE_PARITY_PROPS.to_vec())) else {
                return out;
            };
            if let Some(arr) = res.as_array() {
                out.insert(v.0.to_string(), arr.clone());
            }
        }
        let _ = page.evaluate(
            r#"(function(){ document.documentElement.classList.remove("dark"); document.documentElement.setAttribute("dir","ltr") })()"#,
        );
        out
    };
    let freeze = || {
        let _ = page.evaluate_fn_arg(
            r#"(shell) => { document.body.style.cssText += ";" + shell }"#,
            serde_json::json!(STYLE_PARITY_SHELL),
        );
        let _ = page.add_style_tag(STYLE_PARITY_FREEZE);
        let _ = page.evaluate(
            r#"(function(){ document.getAnimations?.().forEach((a) => a.finish()) })()"#,
        );
        page.wait_for_timeout(120);
    };

    for name in &names {
        let dir = root.join(STYLE_PARITY_DIR).join(name);
        if !dir.join("oracle.html").exists() || !dir.join("shadless.html").exists() {
            continue;
        }
        let def_res = match shell.call(&serde_json::json!({
            "op": "loadContractDef",
            "file": format!(
                "file://{}",
                root.join("tools/contracts/components")
                    .join(format!("{}.mjs", name))
                    .to_string_lossy()
            ),
        })) {
            Ok(r) => r,
            Err(e) => {
                harness_errors.push(format!("{}: harness error — {}", name, first_line(&e)));
                continue;
            }
        };
        let mut def_v = def_res.get("def").cloned().unwrap_or(serde_json::Value::Null);
        crate::emit::css::drop_nulls(&mut def_v);
        let def: SpDef = match serde_json::from_value(def_v) {
            Ok(d) => d,
            Err(e) => {
                harness_errors.push(format!("{}: harness error — {}", name, first_line(&e.to_string())));
                continue;
            }
        };

        let mut style_ignore: std::collections::HashSet<String> = std::collections::HashSet::new();
        for p in &def.style_ignore {
            style_ignore.insert(p.clone());
        }
        for p in STYLE_PARITY_SKIP {
            style_ignore.insert(p.to_string());
        }
        let props: Vec<&str> = STYLE_PARITY_PROPS
            .iter()
            .filter(|p| !style_ignore.contains(**p))
            .copied()
            .collect();

        let mut run_one = || -> Result<(), String> {
            // oracle side (styled by upstream's own oracle.css)
            let abs_o = dir.join("oracle.html");
            page.goto_url(&format!("file://{}", abs_o.to_string_lossy()))?;
            page.wait_for_timeout(400);
            if !def.open.is_empty() {
                page.driver(&def.open)?;
                page.wait_for_timeout(400);
            }
            page.add_style_tag_path(&root.join("build/gates/oracle.css").to_string_lossy())?;
            let _ = page.evaluate(r#"document.documentElement.classList.add("style-nova")"#);
            freeze();
            let oracle_sides = collect_matrix();

            // shadless side (loads its own out.css via relative link)
            let abs_s = dir.join("shadless.html");
            page.goto_url(&format!("file://{}", abs_s.to_string_lossy()))?;
            page.add_style_tag_path(&root.join("dist/out.css").to_string_lossy())?;
            page.wait_for_timeout(400);
            let mut open = def.open_shadless.clone();
            if open.is_empty() {
                open = def.open.clone();
            }
            if !open.is_empty() {
                page.driver(&open)?;
                page.wait_for_timeout(400);
            }
            freeze();
            let shadless_sides = collect_matrix();

            let slot_of = |e: &serde_json::Value| -> String {
                e.get("slot")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string()
            };
            for variant in variants {
                let mut seen_a: std::collections::HashMap<String, usize> =
                    std::collections::HashMap::new();
                let mut seen_b: std::collections::HashMap<String, usize> =
                    std::collections::HashMap::new();
                let key_of = |slot: &str, seen: &mut std::collections::HashMap<String, usize>| {
                    let n = seen.entry(slot.to_string()).or_insert(0);
                    *n += 1;
                    format!("{}#{}", slot, n)
                };
                let mut map_a: std::collections::HashMap<String, std::collections::HashMap<String, String>> =
                    std::collections::HashMap::new();
                let mut order_a: Vec<String> = Vec::new();
                for e in oracle_sides.get(variant.0).into_iter().flatten() {
                    let k = key_of(&slot_of(e), &mut seen_a);
                    let mut sm = std::collections::HashMap::new();
                    if let Some(style) = e.get("style").and_then(|v| v.as_object()) {
                        for (p, v) in style {
                            if let Some(s) = v.as_str() {
                                sm.insert(p.clone(), s.to_string());
                            }
                        }
                    }
                    map_a.insert(k.clone(), sm);
                    order_a.push(k);
                }
                let mut map_b: std::collections::HashMap<String, std::collections::HashMap<String, String>> =
                    std::collections::HashMap::new();
                let mut order_b: Vec<String> = Vec::new();
                for e in shadless_sides.get(variant.0).into_iter().flatten() {
                    let k = key_of(&slot_of(e), &mut seen_b);
                    let mut sm = std::collections::HashMap::new();
                    if let Some(style) = e.get("style").and_then(|v| v.as_object()) {
                        for (p, v) in style {
                            if let Some(s) = v.as_str() {
                                sm.insert(p.clone(), s.to_string());
                            }
                        }
                    }
                    map_b.insert(k.clone(), sm);
                    order_b.push(k);
                }
                let mut push = |k: &str, prop: &str, a: &str, b: &str| {
                    let key = if variant.0 != "light@ltr" {
                        format!("{}@{}", k, variant.0)
                    } else {
                        k.to_string()
                    };
                    cells.push(SpCellEnt {
                        component: name.clone(),
                        key,
                        prop: prop.to_string(),
                        oracle: a.to_string(),
                        shadless: b.to_string(),
                    });
                };
                for k in &order_a {
                    let a = &map_a[k];
                    match map_b.get(k) {
                        None => push(k, "<presence>", "present", "missing"),
                        Some(b) => {
                            for p in &props {
                                let va = parity_norm_value(a.get(*p).map(String::as_str).unwrap_or(""), true);
                                let vb = parity_norm_value(b.get(*p).map(String::as_str).unwrap_or(""), true);
                                if va != vb {
                                    push(k, p, &va, &vb);
                                }
                            }
                        }
                    }
                }
                for k in &order_b {
                    if !map_a.contains_key(k) {
                        push(k, "<presence>", "missing", "present");
                    }
                }
                compared += map_a.len();
            }
            components += 1;
            Ok(())
        };
        if let Err(e) = run_one() {
            harness_errors.push(format!("{}: harness error — {}", name, first_line(&e)));
        }
    }

    if !harness_errors.is_empty() {
        eprintln!("FAIL  style-parity (harness)\n  {}\n", harness_errors.join("\n  "));
        return 1;
    }

    // `flaky` from the raw file (record must work pre-value format too)
    let mut flaky: std::collections::HashSet<String> = std::collections::HashSet::new();
    if let Ok(fb) = std::fs::read_to_string(root.join(STYLE_PARITY_BASELINE)) {
        #[derive(Deserialize)]
        struct Raw {
            #[serde(default)]
            flaky: Vec<String>,
        }
        if let Ok(raw) = serde_json::from_str::<Raw>(&fb) {
            for f in raw.flaky {
                flaky.insert(f);
            }
        }
    }

    let mut ratcheted: Vec<SpCellEnt> = Vec::new();
    for c in &cells {
        if !flaky.contains(&c.component) {
            ratcheted.push(SpCellEnt {
                component: c.component.clone(),
                key: c.key.clone(),
                prop: c.prop.clone(),
                oracle: c.oracle.clone(),
                shadless: c.shadless.clone(),
            });
        }
    }
    let flaky_cells = cells.len() - ratcheted.len();
    let mut actual_cells: Vec<parity_cell> = Vec::new();
    for c in &ratcheted {
        actual_cells.push(parity_cell {
            id: format!("{}/{}/{}", c.component, c.key, c.prop),
            oracle: c.oracle.clone(),
            shadless: c.shadless.clone(),
        });
    }
    let (actual, order) = cell_map(&actual_cells);

    if record || !root.join(STYLE_PARITY_BASELINE).exists() {
        let fl: Vec<String> = flaky.iter().cloned().collect();
        if let Err(e) = write_parity_baseline(
            root,
            STYLE_PARITY_BASELINE,
            "Cells where the shadless fixture's computed style differs from the React oracle, \
with the two values as recorded. This list may only shrink and the values are pinned; \
see the ledger budget style-parity.dirty-cells.",
            &fl,
            &actual,
        ) {
            eprintln!("style-parity: {}", e);
            return 1;
        }
        let comp_set: std::collections::HashSet<&String> =
            ratcheted.iter().map(|c| &c.component).collect();
        println!(
            "style-parity: baseline recorded ({} cells across {} components, {} flaky components excluded)",
            actual.len(),
            comp_set.len(),
            flaky.len()
        );
        return 0;
    }

    if strict && !cells.is_empty() {
        let n = cells.len().min(10);
        let parts: Vec<String> = cells[..n]
            .iter()
            .map(|c| {
                format!(
                    "{}/{}/{}: oracle={} shadless={}",
                    c.component,
                    c.key,
                    c.prop,
                    trunc60(&c.oracle),
                    trunc60(&c.shadless)
                )
            })
            .collect();
        eprintln!(
            "FAIL  style-parity --strict ({} differing cells)\n  {}\n",
            cells.len(),
            parts.join("\n  ")
        );
        return 1;
    }

    let recorded = match load_parity_baseline(root, STYLE_PARITY_BASELINE) {
        Ok(Some((_, r))) => r,
        Ok(None) => Default::default(),
        Err(e) => {
            eprintln!("style-parity: {}", e);
            return 1;
        }
    };
    let d = diff_parity_baseline(&recorded, &actual, &order);
    if !d.appeared.is_empty() {
        let n = d.appeared.len().min(40);
        let parts: Vec<String> = d.appeared[..n]
            .iter()
            .map(|id| format!("{}: {}", id, show_cell(&actual[id])))
            .collect();
        let tail = if d.appeared.len() > 40 {
            format!("\n  … +{} more", d.appeared.len() - 40)
        } else {
            String::new()
        };
        eprintln!(
            "FAIL  style-parity ({} NEW differing cells vs the React oracle)\n  {}{}\n",
            d.appeared.len(),
            parts.join("\n  "),
            tail
        );
        return 1;
    }
    if !d.changed.is_empty() {
        let n = d.changed.len().min(20);
        let parts: Vec<String> = d.changed[..n].iter().map(show_change).collect();
        let tail = if d.changed.len() > 20 {
            format!("\n  … +{} more", d.changed.len() - 20)
        } else {
            String::new()
        };
        eprintln!(
            "FAIL  style-parity ({} recorded cells still differ from the oracle, but by a DIFFERENT amount than what was recorded — look at them again, then re-record)\n  {}{}\n\n  ./build/pipeline style-parity --record\n",
            d.changed.len(),
            parts.join("\n  "),
            tail
        );
        return 1;
    }
    if !d.fixed.is_empty() {
        let n = d.fixed.len().min(12);
        let tail = if d.fixed.len() > 12 {
            format!("\n  … +{} more", d.fixed.len() - 12)
        } else {
            String::new()
        };
        eprintln!(
            "FAIL  style-parity ({} recorded cells no longer differ — record the win so the slack cannot be re-spent)\n  {}{}\n\n  ./build/pipeline style-parity --record && ./build/pipeline ledger --record\n",
            d.fixed.len(),
            d.fixed[..n].join("\n  "),
            tail
        );
        return 1;
    }
    println!(
        "PASS  style-parity ({} components, {} elements compared, {} cells at the recorded baseline incl. their values, {} cells in {} flaky components excluded; --strict is the end state)",
        components,
        compared,
        actual.len(),
        flaky_cells,
        flaky.len()
    );
    shell.close();
    0
}

fn trunc60(s: &str) -> String {
    if s.len() > 60 {
        format!("{}…", &s[..60])
    } else {
        s.to_string()
    }
}

fn first_line(s: &str) -> String {
    match s.find('\n') {
        Some(i) => s[..i].to_string(),
        None => s.to_string(),
    }
}
